### ISSUE TITLE
pyright ignores annotation of variable

### ISSUE THREAD

**Issue by ethframe** [microsoft/pyright#3709]
**Describe the bug**
Pyright ignores the annotation of a variable and narrows its type to the inferred type of expression. This leads to false positives in certain cases.

**To Reproduce**
```
$ pyright bug.py
...
...\bug.py
  ...\bug.py:23:17 - information: Type of "x" is "Foo[U@foo] | Foo[V@foo]"
  ...\bug.py:24:13 - error: Argument of type "Foo[U@foo] | Foo[V@foo]" cannot be assigned 
to parameter "x" of type "Foo[V_co@Bar]" in function "__init__"
    Type "Foo[U@foo] | Foo[V@foo]" cannot be assigned to type "Foo[V_co@Bar]"
      TypeVar "V_co@Foo" is covariant
        Type "U@foo" cannot be assigned to type "V@foo" (reportGeneralTypeIssues)
  ...\bug.py:25:17 - information: Type of "r" is "Bar[Unknown]"
1 error, 0 warnings, 2 informations
```

**Expected behavior**
No errors, type of `x` is `Foo[U@foo | V@foo]`, type of `r` is `Bar[U@foo | V@foo]`, as with mypy:
```
$ mypy bug.py 
bug.py:23: note: Revealed type is "bug.Foo[Union[V`-1, U`-2]]"
bug.py:25: note: Revealed type is "bug.Bar[Union[V`-1, U`-2]]"
Success: no issues found in 1 source file
```

**Screenshots or Code**
```python
from typing import Generic, TypeVar, Union

from typing_extensions import reveal_type

V = TypeVar("V")
V_co = TypeVar("V_co", covariant=True)
U = TypeVar("U")


class Foo(Generic[V_co]):
    pass


class Bar(Generic[V_co]):
    def __init__(self, x: Foo[V_co]):
        pass


def foo(a: Foo[V], b: Foo[U], c: bool) -> Bar[Union[V, U]]:
    x: Foo[Union[V, U]] = a
    if c:
        x = b
    reveal_type(x)
    r = Bar(x)
    reveal_type(r)
    return r
```

**VS Code extension or command-line**
pyright 1.1.260

**Comment by erictraut** [comment]
This is a bug, but the description of the problem isn't quite right. Pyright isn't ignoring the declared type in this case. It is correct for the type to be narrowed on assignment. It's also correct for the narrowed type to be `Foo[U@foo] | Foo[V@foo]` after the `if` block. The problem occurs when a value of this type is passed to the `Bar` constructor. The TypeVar constraint solver is not correctly handling the case where there are two TypeVars involved (`U@foo` and `V@foo`). If one or both of these are concrete types (e.g. `Foo[int] | Foo[str]`), the constraint solver produces the correct result without generating an error.

I'll need to dig into the problem further to understand the underlying cause of the problem with the constraint solver.

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.261, which I just published. It will also be included in an upcoming release of pylance.
